### PR TITLE
Add more extensive envmap supports

### PR DESCRIPTION
**Pull request by panxinmiao**
Try adding more extensive environment map supports, to make it suitable for MeshBasicMaterial, MeshPhongMaterial, and MeshStandardMaterial.

Add more environment map types and mapping modes.
At present, the "cube-reflection" and "cube-refraction" mapping are supported first.

image

The "env_maps.py" example shows the effect of the "cube-reflection" and "cube-refraction" mapping modes of "MeshBasicMaterial", "MeshPhongMaterial", and "MeshStandardMaterial" respectively.

### CODE REVIEW

**Comment by Korijn** [comment]
Status: we need to make time to review this properly but it's not high on our priority list for #392

**Comment by almarklein** [comment]
This is actually pretty self-contained and would be nice to add. It's been sitting still for a while and needs a bit of cleanup though  :) Some general comments:

* Our skybox has been fixed now. 
* I also see some normal flipping. Not sure if we fixed that / soled in a central place since the last change in this PR.
* I think `AbstractMeshMaterial` would be nice name for `_MeshBaseMaterial`.
* Some todo's and commented lines.
*

**Comment by panxinmiao** [comment]
> This is actually pretty self-contained and would be nice to add. No rush though! It's been sitting still for a while and needs a bit of cleanup :) Some general comments:
> 
> * Our skybox has been fixed now.
> * I also see some normal flipping. Not sure if we fixed that / soled in a central place since the last change in this PR.
> * I think `AbstractMeshMaterial` would be nice name for `_MeshBaseMaterial`.
> * Some todo's and commented lines.

Ah,  the unfinished work that I left behind, 😅 

It was the work before #404 and #466. It was during the process of this PR that I discovered the issue with the normals in #404.

Let me see how to update to the current code and complete it, 😅

**Comment by panxinmiao** [comment]
The pygfx library has changed a lot since the start of this PR. I spent some time familiarizing myself with the new API. Now it should be ready for review. 

I'm not sure why the Flake8 check doesn't seem to have passed, but it passed when I checked it locally.

**Comment by almarklein** [comment]
> I'm not sure why the Flake8 check doesn't seem to have passed, but it passed when I checked it locally.

You probably need to `pip install -U black` 😄

**Review by almarklein — COMMENTED** [review]
This is looking good!

The only thing that I am not yet sure about is how to fit this into the class hierarchy of mesh materials.

Right now we have:

* `class MeshBasicMaterial(Material)`: basic, unlit material, doubling as a base class for more special mesh materials.
* `class MeshPhongMaterial(MeshBasicMaterial)`: adds Phong lighting model.
* `class MeshNormalMaterial(MeshBasicMaterial)`: show normals in colors. This ignores many properties of `MeshBasicMaterial`.
* `class MeshNormalLinesMaterial(MeshBasicMaterial)`: same, but showing normals as lines.
* `class MeshSliceMaterial(MeshBasicMaterial)`: Show a slice through the mesh. Ignores some properties of `MeshBasicMaterial`.
* `class MeshStandardMaterial(MeshBasicMaterial)`: PBR rendering. This respects all properties of `MeshBasicMaterial`, though you may wonder if wireframe and flat shading make sense for this material.

My suggestion would be to introduce this new functionality as a new class that is a base class for both Phong and Standard materials:
```
   MeshBasicMaterial
      │
      ├───►  MeshReflectiveMaterial
      │              │
      │              ├───────► MeshPhongMaterial
      │              │
      │              └───────► MeshStandardMaterial
      │
      ├─────►   MeshNormalMaterial
      │
      ├─────►   MeshNormalLinesMaterial
      │
      └─────►   MeshSliceMaterial
``` 

In the example I used "MeshReflectiveMaterial", but perhaps there is a better name.

File: pygfx/renderers/wgpu/meshshader.py [hunk 1]
```diff
@@ -93,6 +93,30 @@ def get_bindings(self, wobject, shared):
...
+                material.env_map, 6
+            )  # TODO: envmap not only cube, but also equirect (hdr format)
+            view = GfxTextureView(material.env_map, view_dim="cube")
+            bindings.append(
+                Binding("s_env_map", "sampler/filtering", sampler, "FRAGMENT")
+            )
+            bindings.append(Binding("t_env_map", "texture/auto", view, "FRAGMENT"))
+
+            if isinstance(material, MeshStandardMaterial):
+                self["use_IBL"] = True
```
[the remark this answers is not in the document]
Comment (panxinmiao) [reply pygfx/renderers/wgpu/meshshader.py:109]: Yes, unlike `MeshBasicMaterial` and `MeshPhongMaterial`, envmap works through "IBL" in `MeshStandardMaterial`.

**Comment by panxinmiao** [comment]
I'm sorry, I recently got COVID-19 and my high fever has only just started to subside.

Regarding class hierarchy of mesh materials, I think the current design of AbstractMeshMaterial and MeshBasicMaterial is reasonable. MeshBasicMaterial provides basic and simple material for most basic objects in simple display scenes. Meanwhile, MeshPhongMaterial and MeshStandardMaterial offer more advanced rendering features. Other materials are more like tools that serve special purposes. They differ from MeshBasicMaterial in many properties, rather than inheritance. They should be parallel with MeshBasicMaterial, both inheriting from AbstractMeshMaterial.

**Comment by almarklein** [comment]
Yes, that would work too, and seems more consistent with how ThreeJS does it 👍  A few requests though:
*  Let's move the wireframe and flat_shading stuff from `AbstractMeshMaterial`  to `MeshBasicMaterial`, since they only apply to "actual" meshes.
* Update the docstrings of both.
* Rename `AbstractMeshMaterial` to `MeshAbstractMaterial` for consistency.
* Expose `MeshAbstractMaterial` in the pygfx main namespace, and docs (via `materials/__init__.py` and `docs/refererence.rst`)

I hope you have a speedy recovery. No rush with this PR! (it's been open for a while, we can wait a bit longer 😉.)

**Comment by panxinmiao** [comment]
Ready for review,  😉.

**Review by almarklein — COMMENTED** [review]
Only a minor tweak to the docs, otherwise LGTM!

**Review by Korijn — APPROVED** [review]

**Review by Korijn — APPROVED** [review]

**Comment by Korijn** [comment]
Seems like we are hitting segmentation faults on RTD... 🤔

**Pull request merged**